### PR TITLE
[mds-web-sockets] First pass at some internal JWT validation.

### PR DESCRIPTION
**Pull request by avatarneil**
Need @invertigo to iron out passing down environment variables so the socket server knows what it's looking for.

## PR Checklist

 - [ ] simple searchable title - `[mds-db] Add PG env var`, `[config] Fix eslint config`
 - [ ] briefly describe the changes in this PR
 - [ ] mark as draft if should not be merged
 - [ ] write tests for all new functionality

## Impacts
- [ ] Provider
- [ ] Agency
- [ ] Audit
- [ ] Policy
- [ ] Compliance
- [ ] Daily
- [ ] Native
- [ ] Policy Author

### CODE REVIEW

File: packages/mds-web-sockets/clients.ts [hunk 1]
```diff
@@ -49,4 +55,30 @@ export class Clients {
...
+    return new Promise(resolve => {
+      // eslint-disable-next-line promise/prefer-await-to-callbacks
+      jwt.verify(token, getKey, { audience: 'ourAudience', issuer: 'ourIssuer' }, (err, decoded) => {
+        if (err) return resolve(false)
+
+        log.info(decoded)
+        return resolve(true)
+      })
+    })
+  }
```
Comment (drtyh2o) [inline packages/mds-web-sockets/clients.ts]: `import { promisify } from 'util'` and get rid of some of these callbacks.

**Comment by macsj200** [comment]
Will it be possible to pull in the relevant issuer and audience from helm/kube?

File: packages/mds-web-sockets/clients.ts [hunk 2]
```diff
@@ -2,12 +2,33 @@ import WebSocket from 'ws'
...
+    /* Technically, this typedef is slightly incorrect, but is to coerce the compiler to happiness without type guarding. One of publicKey or rsaPublicKey *always* exists. */
+    const key: { publicKey?: string; rsaPublicKey?: string } = await promisify(client.getSigningKey)(
+      header.kid ?? 'null'
+    )
+    return key.publicKey || key.rsaPublicKey
+  }
+
+  private static verify: (
+    token: string,
+    secretOrPublicKey: jwt.Secret | GetPublicKeyOrSecret,
```
Comment (BigIdeas) [inline packages/mds-web-sockets/clients.ts]: nit - maybe this arg should be named `publicKeyOrSecret` to match `GetPublicKeyOrSecret`?

File: packages/mds-web-sockets/clients.ts [hunk 3]
```diff
@@ -2,12 +2,33 @@ import WebSocket from 'ws'
...
+import { promisify } from 'util'
 
 export class Clients {
   authenticatedClients: WebSocket[]
 
   subList: { [key: string]: WebSocket[] }
 
+  private static getKey: GetPublicKeyOrSecret = async header => {
+    const client = jwks({
+      jwksUri: 'ourIssuer'
```
Comment (BigIdeas) [inline packages/mds-web-sockets/clients.ts]: Curious if this is the real `jwksUri` we're using or if it will be read in from config in the future

Comment (avatarneil) [reply packages/mds-web-sockets/clients.ts]: Will be read in from config in the future. I need to sync up with alex to see how trivial it is to get all of those variables passed down from the istio layer where they currently live. Same for a bunch of other variables which are just placeholders for now.

File: packages/mds-web-sockets/clients.ts [hunk 4]
```diff
@@ -37,16 +57,41 @@ export class Clients {
...
+
+      const validateAuth = await Clients.checkAuth(token)
+      if (!validateAuth) {
+        client.send(JSON.stringify({ err: new AuthorizationError() }))
+        return
+      }
+
       const scopes = auth?.scope.split(' ') ?? []
+
       if (scopes.includes('admin:all')) {
```
Comment (BigIdeas) [inline packages/mds-web-sockets/clients.ts]: Would the scope to check for here be an argument to this function in the future?

Comment (avatarneil) [reply packages/mds-web-sockets/clients.ts]: Absolutely. Depending on how long we keep this package (there's a chance we might dump it in the future when NATS JetStream comes around), I'll need to spec out scopes for the websocket interface.

**Pull request merged**